### ISSUE TITLE
foldfilter breaks translation from language without spaces to language with spaces

### ISSUE THREAD

**Issue by kpu** [kpu/preprocess#21]
@jelmervdl Problem with foldfilter: if we translate from e.g. fa (without spaces) to en then the output concatenates the last English word of a preceding sentence with the first English word of the following sentence, without an intervening space.  

This impacts quality of zh, ko, fa, and probably other paracrawls.

**Comment by jelmervdl** [comment]
Oh crap… that is not something I had thought about.

By default it just uses the delimiter it chopped out as glue. If there is no delimiter, there is no glue. So no space. It makes sense in the `foldfilter cat` scenario, but not for much for MT…

I'd suggest adding a space between words when the separator was empty (i.e. at a mid-word break instead of some soft line wrapping) except when the `-s` option is given. That option is supposed to pass the delimiters it chops also to the wrapped program. When translating to a language that doesn't do spaces and commas etc. it would make sense to just let the MT system decide to drop them. And `foldfilter -s cat` would still do what you expect.

This should be a simple fix, just an extra if-statement inside https://github.com/kpu/preprocess/blob/master/preprocess/foldfilter_main.cc#L132

**Comment by kpu** [comment]
it's worse than first thought.  

Let's make a `cat` that strips leading and trailing whitespace:
```
#!/usr/bin/env python3
import sys
for l in sys.stdin:
  print(l.strip())
```

```bash
build/bin/foldfilter -w 4 ./remove_space.py <<<"hello hi how are you"
hellohihowareyou
```

**Comment by jelmervdl** [comment]
I misinterpreted my own documentation…

The passing spaces to the wrapped command is the default, [but paracrawl using it *with* the `-s` option](https://github.com/paracrawl/cirrus-scripts/blob/master/04.translate.cpu#L65), which does *not* pass the whitespace to the wrapped program.

With `-s`:
```
foldfilter -s -w 4 ./remove_space.py <<<"hello hi how are you"
hello hi how are you
```

So damage is not that bad (I would have noticed it earlier!) but the original case (input without delimiters) is still a problem. Also maybe the defaults are bad.

**Comment by kpu** [comment]
Yeah ok we should have used `-s` for the fa->en part where I noticed the issue, then thought fa didn't have spaces, then looked at corpora and realized it does have spaces, then blamed ko and zh.  
Default should probably preserve space; not sure what to do about period.
